crates/revm/src/optimism.rs

```diff
@@ -84,30 +91,25 @@ impl TryFrom<&Block> for L1BlockInfo {
 
 impl L1BlockInfo {
     /// Calculate the gas cost of a transaction based on L1 block data posted on L2
-    pub fn calculate_tx_l1_cost(&mut self, tx: &TransactionSigned) -> U256 {
+    pub fn calculate_tx_l1_cost(
+        &mut self,
+        tx: &TransactionSigned,
+    ) -> Result<U256, executor::BlockExecutionError> {
         let rollup_data_gas_cost = U256::from(tx.input().iter().fold(0, |acc, byte| {
             acc + if *byte == 0x00 { ZERO_BYTE_COST } else { NON_ZERO_BYTE_COST }
         }));
 
         if tx.is_deposit() || rollup_data_gas_cost == U256::ZERO {
-            return U256::ZERO
+            return Ok(U256::ZERO)
         }
 
         rollup_data_gas_cost
             .saturating_add(self.l1_fee_overhead)
             .saturating_mul(self.l1_base_fee)
             .saturating_mul(self.l1_fee_scalar)
             .checked_div(U256::from(1_000_000))
-            .unwrap_or_default()
+            .ok_or(executor::BlockExecutionError::L1BlockInfoError {
+                message: "could not calculate tx l1 cost".to_string(),
+            })
```

Comment:
my comment was about `unwrap_or_default` (which was addressed) here as well as `saturating_*` methods. this calculation is protocol-specific, so the behavior at allowed boundaries should be specified, but I'll allow @clabby to weigh in here. again simply not a fan of hidden undefined behavior even if it's highly improbable

Reply:
There is a slight difference in this implementation vs `op-geth`, but it should be safe. `op-geth` uses golang's `big.Int` type, which is signed + has no explicit size:
```golang
// big.Int definition

// A Word represents a single digit of a multi-precision unsigned integer.
type Word uintptr

type nat []Word

type Int struct {
    neg bool // sign
    abs nat  // absolute value of the integer
}
```

It is unrealistic that we'd hit an overflow using unsigned 256 bit integers here I think. We *could* theoretically need the higher bits of precision in the arithmetic for the `add, mul, mul` prior to the division by `1_000_000`, so the result of `(rollup_data_gas_cost + l1_fee_overhead) * l1_base_fee * l1_fee_scalar` could be between `[0,0xf423ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff0bdc1)` with the final result still fitting in `U256`. This would require extremely large input values, though.

@protolambda, are the higher bits of precision ever necessary here?

Reference `op-geth` code: [here](https://github.com/ethereum-optimism/op-geth/blob/39d121af21bc4cb45c46edf2dc1bf33daa28c54d/core/types/rollup_l1_cost.go#L78-L82)

Reply:
256 bit math is safe here.

Reply:
It also just occurred to me the `checked_div` is pointless as it only checks for division by zero, but it is a hardcoded value here so this can't happen. Changing these to regular `add`, `mul`, `div` and adding a comment for safety.